### ISSUE TITLE
Questions about learning procedures in darch.

### ISSUE THREAD

**Issue by shryu8902** [maddin79/darch#6]
Hello.

I'm trying to learn a DBN with darch package version 0.11.9. 
But I'm not sure about my training process is right because I didn't used example codes in the documents.
I trained my DBN with following sequence.

```
## with darch ver 0.11.9
temp <- newDArch(layer,batchSize = batchsize)
temp <- preTrainDArch(temp,train.data,numEpochs = 10)
temp <- fineTuneDArch(temp,train.data,numEpochs=1,bootstrap=F)
exec <- getExecuteFunction(temp)(temp,inputdata)
results <- getExecOutput(exec)
```

So I have 3 questions. 
1) Is it okay to learn DBN with above? 
2) When I first run codes above, I got an error message of 
```
Error in get("matMult", darch.env) : object 'matMult' not found 
```
But when I run same codes after running darch example in documents, error message disappears. 
Is there a parameter that I missed to overcome the error message?
3) Is it okay to use same code with ver 10.0?

Thanks.
Seunghyoung Ryu.

**Comment by saviola777** [comment]
Hello,

1.) Yes, it is possible to train a DBN using your approach. However, as you noticed, there are some (or at least one) parameters missing. That's a mistake on my part, and I will fix this in the future so that it is easier to use darch without the ```darch()``` function that is used in the examples. Additionally, the ```train.data``` that you pass to ```preTrainDArch``` and ```fineTuneDArch``` must be ```DataSet``` objects, not simple ```data.frame```s or similar (see ```createDataSet```).

2.) Please update your darch version to the latest git version (or see the solution in 3.). Previously, I used an environment (darch.env) to store "global" parameters (like matMult). This environment was not properly initialized when manually training the DBN like you did. When you run it with the latest git version, you can simply change your newDArch call as follows:

```
temp <- newDArch(layer,batchSize = batchsize, matMult = `%*%`)
```

As I said, this parameter should be documented in the newDArch function signature, which I will fix soon.

3.) darch 0.10.0 uses the old darch.env object and the same error will happen unless you call darch() once or execute the examples. A workaround would be to use

```
assign("matMult", `%*%`, darch:::darch.env)
```

If there is no special reason for you to do so, I would recommend using the ```darch``` function to create and train your network. Here is the same training process you outlined above, using the ```darch``` and ```predict``` functions (in version 0.10.0 and newer):

```
temp <- darch(train.data, layers = layer, darch.batchSize = batchSize, rbm.numEpochs = 10, darch.numEpochs = 1)
results <- predict(temp, newdata = inputdata)
```

I hope this helps and was not too confusing, I realize that I probably change too many things too quickly sometimes, as I'm not the most experienced programmer, but I hope that I can at some point provide a stable and well document user interface (and feedback / suggestions are always welcome).

If I may ask, why are you not using the ```darch``` function, is there a reason for that? I assumed that a single function offering most parameters which creates and trains the DBN would be easier to use. But if people still want to use the "old" interface, I will try to make it more accessible and/or more similar to how it was before.

**Comment by shryu8902** [comment]
Thank you. It helps me a lot.
I didn't use darch function because I want to shuffle the order of training data which I could obtained better prediction and to check results of training for every iterations. 
So I created shuffled training data and used fineTuneDArch with epoch 1.
And due to the fact that I don't know how to extract error results from darch, I had to calculate mse from prediction results.
These are why I'm using separate functions.

**Comment by saviola777** [comment]
I see. In recent git versions, shuffling of training data (introduced in cab92e1) is done automatically, and the results (i.e. MSE and classification error) of training for every iteration can be extracted using ```darch@stats``` which contains lists of errors, times needed etc. for each epoch (will integrate these stats into the ```print.DArch``` function in the future). If you still feel like training one epoch at a time, you can do so using the ```darch``` function as well:

```
darch <- NULL

for (i in 1:n)
{
  # pass existing DArch instance to continue training
  darch <- darch(train.data, […], darch = darch)
  # here you can extract information from the DArch instance
}
```

**Comment by saviola777** [comment]
Closing after more than a month of inactivity, please create a new issue if something related comes up.

**Comment by kuldeepldec** [comment]
Hi saviola I have made dataSet after doing preTrainDArch,in fineTuneDArch i am getting error of Error in if (!getDropoutOneMaskPerEpoch(darch)) { : argument is of length zero.

**Comment by kuldeepldec** [comment]
I have used mod = fineTuneDArch(darch = dbn,dataSet = dataSet,numEpochs = 10,bootstrap = T)

**Comment by saviola777** [comment]
Hi kuldeepldec,

please try to use the ```darch``` function instead of ```preTrainDArch``` and ```fineTuneDArch```. If the problem perists, please open a new issue and post your complete code and error output.

**Comment by kuldeepldec** [comment]
Hi saviloa thank you for your reply, darch is working correctly. I tried to use pretrainDarch before it was not working. This time i got success on pretrainDarch but fineTuneDarch is not working

**Comment by saviola777** [comment]
Sorry, the functions ```preTrainDArch``` and fineTuneDArch``` are no longer supported by darch, they are only used internally. Please use ```darch()``` instead, it contains both ```preTrainDArch``` and fineTuneDArch```. If you need help converting your code to the new interface, please provide the full code.
